src/primitives/transaction.h

```diff
@@ -607,32 +822,66 @@ class CTransaction
             nVersionGroupId == ZFUTURE_VERSION_GROUP_ID &&
             nVersion == ZFUTURE_TX_VERSION;
 
-        if (fOverwintered && !(isOverwinterV3 || isSaplingV4 || isFuture)) {
+        if (fOverwintered && !(isOverwinterV3 || isSaplingV4 || isZip225V5 || isFuture)) {
             throw std::ios_base::failure("Unknown transaction format");
         }
 
-        READWRITE(*const_cast<std::vector<CTxIn>*>(&vin));
-        READWRITE(*const_cast<std::vector<CTxOut>*>(&vout));
-        READWRITE(*const_cast<uint32_t*>(&nLockTime));
-        if (isOverwinterV3 || isSaplingV4 || isFuture) {
+        if (isZip225V5) {
+            // Common Transaction Fields (plus version bytes above)
+            READWRITE(*const_cast<uint32_t*>(&this->nConsensusBranchId));
+            READWRITE(*const_cast<uint32_t*>(&nLockTime));
             READWRITE(*const_cast<uint32_t*>(&nExpiryHeight));
-        }
-        if (isSaplingV4 || isFuture) {
-            READWRITE(*const_cast<CAmount*>(&valueBalance));
-            READWRITE(*const_cast<std::vector<SpendDescription>*>(&vShieldedSpend));
-            READWRITE(*const_cast<std::vector<OutputDescription>*>(&vShieldedOutput));
-        }
-        if (nVersion >= 2) {
-            // These fields do not depend on fOverwintered
-            auto os = WithVersion(&s, static_cast<int>(header));
-            ::SerReadWrite(os, *const_cast<std::vector<JSDescription>*>(&vJoinSplit), ser_action);
-            if (vJoinSplit.size() > 0) {
-                READWRITE(*const_cast<Ed25519VerificationKey*>(&joinSplitPubKey));
-                READWRITE(*const_cast<Ed25519Signature*>(&joinSplitSig));
+
+            // Transparent Transaction Fields
+            READWRITE(*const_cast<std::vector<CTxIn>*>(&vin));
+            READWRITE(*const_cast<std::vector<CTxOut>*>(&vout));
+
+            // Sapling Transaction Fields
+            if (ser_action.ForRead()) {
+                SaplingBundle saplingBundle;
+                READWRITE(saplingBundle);
+                *const_cast<std::vector<SpendDescription>*>(&vShieldedSpend) =
```

Comment:
This is undefined behaviour. I know it was undefined behaviour before this PR, but every time we change this code, we have a new chance for it to be miscompiled. It should have been fixed long ago (#967).

Reply:
I don't think we can fix this in this PR. However, given that for Orchard we are parsing everything in Rust and then using calls to Rust to inspect Orchard data, we require accessors for those, and so we can make a start towards fixing this by doing the same for all new variables going forward.

I will move `nConsensusBranchId` and `orchardBundle` to be non-const private member variables.